pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,998 @@
+	fn calculate_weight(
+		curve_type: WeightCurveType,
+		a_x: BlockNumber,
+		b_x: BlockNumber,
+		a_y: LBPWeight,
+		b_y: LBPWeight,
+		at: BlockNumber,
+	) -> Result<LBPWeight, ()>;
+}
+
+pub struct LBPWeightFunction;
+impl<BlockNumber: AtLeast32BitUnsigned> LBPWeightCalculation<BlockNumber> for LBPWeightFunction {
+	fn calculate_weight(
+		curve_type: WeightCurveType,
+		a_x: BlockNumber,
+		b_x: BlockNumber,
+		a_y: LBPWeight,
+		b_y: LBPWeight,
+		at: BlockNumber,
+	) -> Result<LBPWeight, ()> {
+		match curve_type {
+			WeightCurveType::Linear => {
+				hydra_dx_math::lbp::calculate_linear_weights(a_x, b_x, a_y, b_y, at).map_err(|_| ())
+			}
+			WeightCurveType::Constant => Ok(a_y),
+		}
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::OriginFor;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Multi currency for transfer of currencies
+		type MultiCurrency: MultiCurrencyExtended<Self::AccountId, CurrencyId = AssetId, Amount = Amount, Balance = Balance>
+			+ MultiReservableCurrency<Self::AccountId>;
+
+		#[pallet::constant]
+		/// Native Asset Id
+		type NativeAssetId: Get<AssetId>;
+
+		/// The origin which can create a new pool
+		type CreatePoolOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Function for calculation of LBP weights
+		type LBPWeightFunction: LBPWeightCalculation<Self::BlockNumber>;
+
+		/// Mapping of asset pairs to unique pool identities
+		type AssetPairPoolId: AssetPairPoolIdFor<AssetId, PoolId<Self>>;
+
+		#[pallet::constant]
+		/// Storage fee
+		type PoolDeposit: Get<BalanceOf<Self>>;
+
+		/// Trading fee rate
+		type ExchangeFee: Get<fee::Fee>;
+
+		/// Weight information for the extrinsics
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Pool assets can not be the same
+		CannotCreatePoolWithSameAssets,
+		/// Initial liquidity should be non-zero
+		CannotCreatePoolWithZeroLiquidity,
+		/// Account is not a pool owner
+		NotOwner,
+		/// Sale already started
+		SaleStarted,
+		/// Sale is still in progress
+		SaleNotEnded,
+		/// Sale is not running
+		SaleIsNotRunning,
+		/// Provided data are not valid
+		InvalidData,
+		/// Sale already ended
+		CannotPauseEndedPool,
+		/// Sale already ended
+		CannotUnpauseEndedPool,
+		/// Sale is already paused
+		CannotPausePausedPool,
+		/// Pool is set to be not pausable
+		PoolIsNotPausable,
+		/// Pool is not paused
+		PoolIsNotPaused,
+		/// Sale duration is too long
+		MaxSaleDurationExceeded,
+		/// Liquidity being added should not be zero
+		CannotAddZeroLiquidity,
+		/// Can not remove zero liquidity
+		CannotRemoveZeroLiquidity,
+		/// Overflow after adding liquidity
+		LiquidityOverflow,
```

Comment:
No tests.

Reply:
I have decided to remove the LiquidityOverflow error because I was not able to create a test for it (we can't simulate such a scenario)